File: 프로그래머스/unrated/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

class Solution {
    
    private static int N, M, answer;
    private static boolean[][] visit;
    private static int[] A;
    private static final int[][] dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    
    public int solution(int[][] land) {
        N = land.length;
        M = land[0].length;
        
        A = new int[M];
        visit = new boolean[N][M];
        
        for(int i = 0; i < M; i++) {
            for(int j = 0; j < N; j++) {
                if(!visit[j][i] && land[j][i] == 1) {
                    bfs(j, i, land);
                }
            }
        }
        
        for(int i = 0; i < M; i++) {
            answer = Math.max(answer, A[i]);
        }
        
        return answer;
    }
    
    private static void bfs(int x, int y, int[][] land) {
        int count = 1;
        
        boolean[] used = new boolean[M];
        
        Queue<Integer> q = new LinkedList<>();
        visit[x][y] = true;
        used[y] = true;
        q.add(x);
        q.add(y);
        
        while(!q.isEmpty()) {
            x = q.poll();
            y = q.poll();
            
            for(int k = 0; k < 4; k++) {
                int nx = x + dirs[k][0];
                int ny = y + dirs[k][1];
                
                if(nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
                if(visit[nx][ny]) continue;
                if(land[nx][ny] == 0) continue;
                
                visit[nx][ny] = true;
                used[ny] = true;
                count++;
                q.add(nx);
                q.add(ny);
            }
        }
        
        for(int i = 0; i < M; i++) {
            if(used[i]) {
                A[i] += count;
            }
        }
    }
    
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
class Queue {
    constructor() {
        this.queue = [];
    }
    
    enqueue(n) {
        this.queue.push(n);
    }
    
    dequeue() {
        return this.queue.shift();
    }
    
    isEmpty() {
        return this.queue.length === 0;
    }
}

let n, m;
let visited = [];
let total = [];
const dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]];

function solution(land) {
    let answer = 0;
    
    n = land.length;
    m = land[0].length;
    
    for(let i = 0; i < n; i++) {
        visited.push([]);
        for(let j = 0; j < m; j++) {
            visited[i].push(false);
        }
    }
    
    for(let i = 0; i < m; i++) {
        total[i] = 0;
    }
    
    for(let i = 0; i < n; i++) {
        for(let j = 0; j < m; j++) {
            if(visited[i][j]) continue;
            if(land[i][j] === 0) continue;
            
            bfs(i, j, land);
        }
    }
    
    for(let amount of total) {
        answer = Math.max(amount, answer);
    }
    
    return answer;
}

const bfs = (a, b, land) => {
    const q = new Queue();
    let visitHole = [];
    visited[a][b] = true;
    q.enqueue([a, b]);
    visitHole.push(b);
    
    let amount = 0;
    
    while(!q.isEmpty()) {
        const [x, y] = q.dequeue();
        amount++;
        
        for(let dir of dirs) {
            const [a, b] = dir;
            
            const nx = x + a;
            const ny = y + b;
            
            if(nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
            if(land[nx][ny] === 0) continue;
            if(visited[nx][ny]) continue;
            
            visited[nx][ny] = true;
            if(!visitHole.includes(ny)) visitHole.push(ny);
            q.enqueue([nx, ny]);
            
        }
    }
    
    for(let idx of visitHole) {
        total[idx] += amount;
    }
}